pennylane/hf/tapering.py

```diff
@@ -0,0 +1,290 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the functions needed for tapering qubits using symmetries.
+"""
+from copy import deepcopy
+import numpy as np
+import pennylane as qml
+
+
+def _binary_matrix(terms, num_qubits):
+    r"""Get a binary matrix representation of the Hamiltonian where each row corresponds to a
+    Pauli term, which is represented by a concatenation of Z and X vectors.
+
+    Args:
+        terms (Iterable[pennylane.Observable]): operators defining the Hamiltonian
+        num_qubits (int): number of wires required to define the Hamiltonian
+    Returns:
+        array[int]: binary matrix representation of the Hamiltonian of shape
+        :math:`len(terms) \times 2*num_qubits`
+
+    **Example**
+
+    .. code-block:: python
+
+        >>> terms = [qml.PauliZ(wires=[0]) @ qml.PauliX(wires=[1]),
+        ...          qml.PauliZ(wires=[0]) @ qml.PauliY(wires=[2]),
+        ...          qml.PauliX(wires=[0]) @ qml.PauliY(wires=[3])]
+        >>> _binary_matrix(terms, 4)
+         array([[1, 0, 0, 0, 0, 1, 0, 0],
+                [1, 0, 1, 0, 0, 0, 1, 0],
+                [0, 0, 0, 1, 1, 0, 0, 1]]))
+
+    """
+
+    binary_matrix = np.zeros((len(terms), 2 * num_qubits), dtype=int)
+    for idx, term in enumerate(terms):
+        ops, wires = term.name, term.wires
+        if len(term.wires) == 1:
+            ops = [ops]
+        for op, wire in zip(ops, wires):
+            if op in ["PauliX", "PauliY"]:
+                binary_matrix[idx][wire + num_qubits] = 1
+            if op in ["PauliZ", "PauliY"]:
+                binary_matrix[idx][wire] = 1
+
+    return binary_matrix
+
+
+def _reduced_row_echelon(binary_matrix):
+    r"""Returns the reduced row echelon form (RREF) of a matrix in a binary finite field :math:`\mathbb{Z}_2`.
+
+    Args:
+        binary_matrix (array[int]): binary matrix representation of the Hamiltonian
+    Returns:
+        array[int]: reduced row-echelon form of the given `binary_matrix`
+
+    **Example**
+
+    .. code-block:: python
+
+        >>> binary_matrix = np.array([[1, 0, 0, 0, 0, 1, 0, 0],
+        ...                           [1, 0, 1, 0, 0, 0, 1, 0],
+        ...                           [0, 0, 0, 1, 1, 0, 0, 1]])
+        >>> _reduced_row_echelon(binary_matrix)
+         array([[1, 0, 0, 0, 0, 1, 0, 0],
+                [0, 0, 1, 1, 1, 1, 1, 1],
+                [0, 0, 0, 1, 1, 0, 0, 1]])
```

Comment:
Is this actually the correct rref for the given matrix? I was under the impression that in rref the leading 1s should be the only non zero entry in their respective columns. This could just me not knowing what rref is for binary matrices, but I would guess that the rref is supposed to be: 

```
array([[1, 0, 0, 0, 0, 1, 0, 0],
       [0, 0, 1, 0, 0, 1, 1, 0],
       [0, 0, 0, 1, 1, 0, 0, 1]])
```

Reply:
Good catch! You're absolutely correct. I have fixed this now.